core/src/main/java/io/parsingdata/metal/data/selection/ByOffset.java

```diff
@@ -31,35 +34,39 @@
     }
 
-    public static ParseItem findItemAtOffset(final ImmutableList<ParseItem> items, final long offset, final Source source) {
+    public static SafeTrampoline<ParseItem> findItemAtOffset(final ImmutableList<ParseItem> items, final long offset, final Source source) {
         checkNotNull(items, "items");
         checkNotNull(source, "source");
-        if (items.isEmpty()) { return null; }
+        if (items.isEmpty()) { return complete(() -> null); }
         final ParseItem head = items.head;
-        if (head.isValue() && matchesLocation(head.asValue(), offset, source)) { return head; }
+        if (head.isValue() && matchesLocation(head.asValue(), offset, source)) { return complete(() -> head); }
         if (head.isGraph()) {
-            final ParseValue value = getLowestOffsetValue(head.asGraph(), null);
-            if (value != null && matchesLocation(value, offset, source)) { return head; }
+            final ParseValue value = getLowestOffsetValue(ImmutableList.create(head.asGraph()), null).computeResult();
+            if (value != null && matchesLocation(value, offset, source)) { return complete(() -> head); }
         }
-        return findItemAtOffset(items.tail, offset, source);
+        return intermediate(() -> findItemAtOffset(items.tail, offset, source));
     }
 
     private static boolean matchesLocation(final ParseValue value, final long offset, final Source source) {
         return value.slice.offset == offset && value.slice.source.equals(source);
     }
 
-    private static ParseValue getLowestOffsetValue(final ParseGraph graph, final ParseValue lowest) {
-        if (graph.isEmpty() || !graph.getDefinition().isLocal()) { return lowest; }
+    private static SafeTrampoline<ParseValue> getLowestOffsetValue(final ImmutableList<ParseGraph> graphList, final ParseValue lowest) {
+        if (graphList.isEmpty()) { return complete(() -> lowest); }
+        final ParseGraph graph = graphList.head;
+        if (graph.isEmpty() || !graph.getDefinition().isLocal()) {
+            return intermediate(() -> getLowestOffsetValue(graphList.tail, lowest));
+        }
         if (graph.head.isValue()) {
-            return getLowestOffsetValue(graph.tail, lowest == null || lowest.slice.offset > graph.head.asValue().slice.offset ? graph.head.asValue() : lowest);
+            return intermediate(() -> getLowestOffsetValue(graphList.tail.add(graph.tail), lowest == null || lowest.slice.offset > graph.head.asValue().slice.offset ? graph.head.asValue() : lowest));
         }
         if (graph.head.isGraph()) {
-            return getLowestOffsetValue(graph.tail, getLowestOffsetValue(graph.head.asGraph(), lowest));
+            return intermediate(() -> getLowestOffsetValue(graphList.tail.add(graph.head.asGraph()).add(graph.tail), lowest));
         }
-        return getLowestOffsetValue(graph.tail, lowest);
+        return intermediate(() -> getLowestOffsetValue(graphList.tail.add(graph.tail), lowest));
```

Comment:
Agreed. I think instead of refactoring many methods to a similar pattern, we should postpone that and do it as part of #172.